kernel/comps/block/src/bio.rs

```diff
@@ -361,88 +365,351 @@ pub enum BioStatus {
     IoError = 5,
 }
 
-/// `BioSegment` is a smallest memory unit in block I/O.
-///
-/// It is a contiguous memory region that contains multiple sectors.
+/// `BioSegment` is the basic memory unit of a block I/O request.
 #[derive(Debug, Clone)]
 pub struct BioSegment {
-    /// The contiguous pages on which this segment resides.
-    pages: SegmentSlice,
-    /// The starting offset (in bytes) within the first page.
-    /// The offset should always be aligned to the sector size and
-    /// must not exceed the size of a single page.
-    offset: AlignedUsize<SECTOR_SIZE>,
-    /// The total length (in bytes).
-    /// The length can span multiple pages and should be aligned to
-    /// the sector size.
-    len: AlignedUsize<SECTOR_SIZE>,
+    inner: Arc<BioSegmentInner>,
+}
+
+/// The inner part of `BioSegment`.
+#[derive(Debug)]
+struct BioSegmentInner {
+    /// Internal DMA slice.
+    dma_slice: DmaStreamSlice,
+    /// Whether the segment is allocated from the pool.
+    from_pool: bool,
```

Comment:
The primary purpose of using `Arc<BioSegmentInner>` is we need to find the correct timing to return the segment to the pool, which is in `BioSegmentInner::drop()`. Since the drop can happen at a different place&time, e.g., at the end of irq handler of virtio-blk. So this requires a reference counter to track, and to avoid double free.